x/twap/listeners_test.go

```diff
@@ -77,58 +78,222 @@ func (s *TestSuite) TestAfterPoolCreatedHook() {
+func (s *TestSuite) TestAfterEpochEnd() {
+	s.SetupTest()
 	s.Ctx = s.Ctx.WithBlockTime(baseTime)
-
 	poolId := s.PrepareBalancerPoolWithCoins(defaultUniV2Coins...)
-	expectedHistoricalTwap, err := types.NewTwapRecord(s.App.GAMMKeeper, s.Ctx, poolId, denom0, denom1)
+	twapBeforeEpoch, err := s.twapkeeper.GetRecordAtOrBeforeTime(s.Ctx, poolId, s.Ctx.BlockTime(), denom0, denom1)
 	s.Require().NoError(err)
+	fmt.Println(twapBeforeEpoch.Time.String())
 
-	s.EndBlock()
-	s.Commit() // clear transient store
-	// Now on a clean state after a create pool
-	s.Require().Equal(baseTime.Add(time.Second), s.Ctx.BlockTime())
-	s.RunBasicSwap(poolId)
+	pruneEpochIdentifier := s.App.TwapKeeper.PruneEpochIdentifier(s.Ctx)
+	recordHistoryKeepPeriod := 48 * time.Hour
 
-	// accumulators are default right here
-	expectedLatestTwapUpToAccum, err := types.NewTwapRecord(s.App.GAMMKeeper, s.Ctx, poolId, denom0, denom1)
-	s.Require().NoError(err)
-	// ensure different spot prices
-	s.Require().NotEqual(expectedHistoricalTwap.P0LastSpotPrice, expectedLatestTwapUpToAccum.P0LastSpotPrice)
-	s.Require().NotEqual(expectedHistoricalTwap.P1LastSpotPrice, expectedLatestTwapUpToAccum.P1LastSpotPrice)
+	// make prune record time pass by, running prune epoch after this should prune old record
+	s.Ctx = s.Ctx.WithBlockTime(baseTime.Add(recordHistoryKeepPeriod).Add(time.Second))
 
-	s.EndBlock()
+	allEpochs := s.App.EpochsKeeper.AllEpochInfos(s.Ctx)
+	// iterate through all epoch, ensure that epoch only gets pruned in prune epoch identifier
+	for _, epoch := range allEpochs {
+		s.App.TwapKeeper.EpochHooks().AfterEpochEnd(s.Ctx, epoch.Identifier, int64(1))
 
-	// check records
-	historicalOldTwap, err := s.twapkeeper.GetRecordAtOrBeforeTime(s.Ctx, poolId, baseTime, denom0, denom1)
-	s.Require().NoError(err)
-	s.Require().Equal(expectedHistoricalTwap, historicalOldTwap)
+		recentTwapRecords, err := s.twapkeeper.GetRecordAtOrBeforeTime(s.Ctx, poolId, baseTime.Add(epoch.Duration), denom0, denom1)
 
-	latestTwap, err := s.twapkeeper.GetMostRecentRecordStoreRepresentation(s.Ctx, poolId, denom0, denom1)
-	s.Require().NoError(err)
-	s.Require().Equal(latestTwap.P0LastSpotPrice, expectedLatestTwapUpToAccum.P0LastSpotPrice)
-	s.Require().Equal(latestTwap.P1LastSpotPrice, expectedLatestTwapUpToAccum.P1LastSpotPrice)
+		// old record should have been pruned here
+		if epoch.Identifier == pruneEpochIdentifier {
+			s.Require().Error(err)
+			s.Require().NotEqual(twapBeforeEpoch, recentTwapRecords)
 
-	latestTwap2, err := s.twapkeeper.GetRecordAtOrBeforeTime(s.Ctx, poolId, s.Ctx.BlockTime(), denom0, denom1)
-	s.Require().NoError(err)
-	s.Require().Equal(latestTwap, latestTwap2)
+			// quit test once the record has been pruned
+			return
+		} else { // pruning should not be triggered at first, not pruning epoch
+			s.Require().NoError(err)
+			s.Require().Equal(twapBeforeEpoch, recentTwapRecords)
```

Comment:
I don't think we are ever hitting this

Reply:
Good catch! solvefd this by reverse iterating range of epoch identifiers